### PR TITLE
Merge Questionnaire API

### PR DESCRIPTION
**Pull request by aaronzsun**

### CODE REVIEW

**Comment by plynchnlm** [comment]
These tests are failing.  I checked out your branch, ran npm ci && npm run build && npm run test, and got:
✔ Browser application bundle generation complete.

Error: projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:21:69 - error TS2339: Property 'to' does not exist on type 'Matchers<object>'.

21       expect(service.addTotalScoreRule(fhir.phq9, phq9ScoreLinkId)).to.include(phq9array[i]);
                                                                       ~~
projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:26:67 - error TS2339: Property 'to' does not exist on type 'Matchers<object>'.

26     expect(service.addTotalScoreRule(fhir.phq9, phq9ScoreLinkId)).to.include(phq9Val);
                                                                     ~~
projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:30:67 - error TS2339: Property 'to' does not exist on type 'Matchers<object>'.

30     expect(service.addTotalScoreRule(fhir.phq9, phq9ScoreLinkId)).to.include(phq9Score);
                                                                     ~~

**Comment by plynchnlm** [comment]
Error: projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:21:69 - error TS2339: Property 'to' does not exist on type 'Matchers<object>'.

21       expect(service.addTotalScoreRule(fhir.phq9, phq9ScoreLinkId)).to.include(phq9array[i]);
                                                                       ~~
projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:26:67 - error TS2339: Property 'to' does not exist on type 'Matchers<object>'.

26     expect(service.addTotalScoreRule(fhir.phq9, phq9ScoreLinkId)).to.include(phq9Val);
                                                                     ~~
projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:30:67 - error TS2339: Property 'to' does not exist on type 'Matchers<object>'.

30     expect(service.addTotalScoreRule(fhir.phq9, phq9ScoreLinkId)).to.include(phq9Score);
                                                                     ~~

File: src/app/app.component.ts [hunk 1]
```diff
@@ -11,15 +11,19 @@ export class AppComponent {
   fhirPreview: string;
   questionnaire = 'bmi';
   fhir = fhir;
-  linkId = context;
+  phq9ScoreLinkId = phq9ScoreLinkId;
```
Comment (mazilumt) [inline src/app/app.component.ts:14]: This is not only specific to the phq9. Also removing linkId makes the e2e tests fail.

File: src/app/app.component.ts [hunk 2]
```diff
@@ -11,15 +11,19 @@ export class AppComponent {
...
   onChange(): void {
     this.fhirPreview = '';
   }
 
   onSave(fhirResult): void {
     this.fhirPreview = JSON.stringify(fhirResult, null, 2);
   }
+
+  addTotalScoreRule(): void {
+    const result = this.ruleEditorService.addTotalScoreRule(fhir.phq9, phq9ScoreLinkId);
```
Comment (mazilumt) [inline src/app/app.component.ts:27]: What are you doing with the result? You don't really have to add this function here, you can just keep the unit test for the functionality.

File: src/app/app.component.ts [hunk 3]
```diff
@@ -11,15 +11,19 @@ export class AppComponent {
   fhirPreview: string;
   questionnaire = 'bmi';
   fhir = fhir;
-  linkId = context;
+  phq9ScoreLinkId = phq9ScoreLinkId;
 
   constructor(private ruleEditorService: RuleEditorService) { }
-
+s
```
Comment (mazilumt) [inline src/app/app.component.ts:17]: This should not be here.

File: src/app/app.component.html [hunk 4]
```diff
@@ -7,6 +7,9 @@ <h1>Demo Application</h1>
     <option value="bmi">BMI Calculation</option>
     <option value="phq9">PHQ9 (no FHIRPath)</option>
   </select>
+  <button (click)="addTotalScoreRule()">
```
Comment (mazilumt) [inline src/app/app.component.html:10]: Like the comment below this doesn't really have to be exposed in the UI.

File: projects/ng-rule-editor/tsconfig.spec.json [hunk 5]
```diff
@@ -3,8 +3,13 @@
   "extends": "../../tsconfig.json",
   "compilerOptions": {
     "outDir": "../../out-tsc/spec",
+    "moduleResolution": "node",
     "types": [
       "jasmine"
+    ],
+    "typeRoots": [
+      "node_modules/@types",
+      "node_modules/@types/node"
```
Comment (mazilumt) [inline projects/ng-rule-editor/tsconfig.spec.json:12]: Can you revert the changes back please, it was meant for debugging.

File: projects/ng-rule-editor/src/lib/rule-editor.service.ts [hunk 6]
```diff
@@ -390,11 +390,24 @@ export class RuleEditorService {
...
+    this.linkIdContext = linkId;
+    return this.exportSumOfScores();
+  }
+
   /**
    * Given the current FHIR questionnaire definition and a linkId return a new FHIR
    * Questionnaire with a calculated expression at the given linkId which sums up
    * all the ordinal values in the questionnaire
    */
+
```
Comment (mazilumt) [inline projects/ng-rule-editor/src/lib/rule-editor.service.ts:410]: I think it makes more sense not to have a newline here.

File: src/app/mock-data.ts [hunk 7]
```diff
@@ -1387,3 +1387,46 @@ export const fhir = {
   ]
 }
 };
+
```
Comment (mazilumt) [inline src/app/mock-data.ts:1390]: I think most of these are already in the spec file and should not be here since this file is going away.

**Review by mazilumt — CHANGES_REQUESTED** [review]

File: src/app/app.component.ts (hunk 2 above)
Comment (aaronzsun) [reply src/app/app.component.ts:27]: I think it was to get the Calculate Total Scores button to work. Should I set it equal to this.fhirPreview instead?

File: src/app/app.component.ts (hunk 3 above)
Comment (aaronzsun) [reply src/app/app.component.ts:17]: Fixed

File: src/app/app.component.ts (hunk 1 above)
Comment (aaronzsun) [reply src/app/app.component.ts:14]: Fixed

File: projects/ng-rule-editor/tsconfig.spec.json (hunk 5 above)
Comment (aaronzsun) [reply projects/ng-rule-editor/tsconfig.spec.json:12]: Fixed

File: src/app/app.component.html (hunk 4 above)
Comment (aaronzsun) [reply src/app/app.component.html:10]: Fixed

File: projects/ng-rule-editor/src/lib/rule-editor.service.ts (hunk 6 above)
Comment (aaronzsun) [reply projects/ng-rule-editor/src/lib/rule-editor.service.ts:410]: Fixed

File: src/app/mock-data.ts (hunk 7 above)
Comment (aaronzsun) [reply src/app/mock-data.ts:1390]: Fixed

File: src/app/app.component.ts (hunk 2 above)
Comment (mazilumt) [reply src/app/app.component.ts:27]: I don't think we need the button on the UI.

Comment (aaronzsun) [reply src/app/app.component.ts:27]: Fixed, I commented it out but can also just remove it if you'd like.

Comment (mazilumt) [reply src/app/app.component.ts:27]: Yes, please remove don't comment it out. It's available in git history if we need it back.

File: projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts [hunk 8]
```diff
@@ -2,15 +2,54 @@ import { TestBed } from '@angular/core/testing';
...
+const outputItem = {
+  "url": "http://hl7.org/fhir/StructureDefinition/variable",
+  "valueExpression": {
+    "name": "i",
+    "language": "text/fhirpath",
+    "expression": "%questionnaire.item.where(linkId = '/44260-8').answerOption.where(valueCoding.code=%resource.item.where(linkId = '/44260-8').answer.valueCoding.code).extension.where(url='http://hl7.org/fhir/StructureDefinition/ordinalValue').value"
+  }
+};
+
+// @ts-ignore
```
Comment (mazilumt) [inline projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:25]: Can you also remove these 3 lines, they're not used.

File: projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts [hunk 9]
```diff
@@ -2,15 +2,54 @@ import { TestBed } from '@angular/core/testing';
...
+
+  });
+
+  it("should return scored questionnaire with var10", () => {
+
+    // @ts-ignore
+    expect(output.item[9].extension[12]).toEqual(outputItem);
+  });
 });
+import { from } from 'rxjs';
```
Comment (mazilumt) [inline projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:55]: I also missed this line.

File: projects/ng-rule-editor/tsconfig.spec.json [hunk 10]
```diff
@@ -3,6 +3,7 @@
   "extends": "../../tsconfig.json",
   "compilerOptions": {
     "outDir": "../../out-tsc/spec",
+    "moduleResolution": "node",
```
Comment (mazilumt) [inline projects/ng-rule-editor/tsconfig.spec.json:6]: Remove this line too.

File: src/app/app.component.html [hunk 11]
```diff
@@ -7,6 +7,9 @@ <h1>Demo Application</h1>
     <option value="bmi">BMI Calculation</option>
     <option value="phq9">PHQ9 (no FHIRPath)</option>
   </select>
+ <!-- <button (click)="addTotalScoreRule()">
```
Comment (mazilumt) [inline src/app/app.component.html:10]: Remove, don't uncomment please.

File: src/app/app.component.ts (hunk 1 above)
Comment (mazilumt) [reply src/app/app.component.ts:15]: Also what does the variable phq9ScoreLinkId do? (Not the one in mock-data)

File: projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts (hunk 8 above)
Comment (aaronzsun) [reply projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:25]: Fixed

File: projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts (hunk 9 above)
Comment (aaronzsun) [reply projects/ng-rule-editor/src/lib/rule-editor.service.spec.ts:55]: Fixed

File: projects/ng-rule-editor/tsconfig.spec.json (hunk 10 above)
Comment (aaronzsun) [reply projects/ng-rule-editor/tsconfig.spec.json:6]: Fixed

File: src/app/app.component.html (hunk 11 above)
Comment (aaronzsun) [reply src/app/app.component.html:10]: Fixed

**Review by plynchnlm — CHANGES_REQUESTED** [review]
The tests are failing because "npm run test" just runs ng e2e which updates webdriver to an incompatible version.  We discussed this problem on Teams today.  The solution is to revise the "npm run test" script so that it updates webdriver to a compatible version, and so that ng e2e does not update webdriver.  In the "test" script in package.json, replace "ng e2e" with "node ./node_modules/.bin/webdriver-manager update --versions.chrome  $(google-chrome --version |awk '{print $3}') && ng e2e --webdriver-update=false".  (I ran that and the tests passed.)

**Review by plynchnlm — APPROVED** [review]

**Review by mazilumt — APPROVED** [review]
